### ISSUE TITLE
Keras Metric Multiple Outputs / Inputs

### ISSUE THREAD

**Issue by AmitMY** [tensorflow/tensorflow#41695]
**System information**
- TensorFlow version (you are using): 2.2.0
- Are you willing to contribute it (Yes/No): No

**Describe the feature and the current behavior/state.**
Keras custom `metric` should optionally get as input all inputs and all outputs of the same batch.
This will allow creating metrics that are dependent on 2 or more outputs, or dependent on the input, etc.

**Will this change the current api? How?**
Yes, a metric will have a property `use_all_inputs` or something like that

**Who will benefit with this feature?**
Anyone who is looking into custom metrics.

**Any Other info.**
It is possible to do something similar with a callback, but that requires running `predict` twice per sample per epoch, and does not work for `model.evaluate`.

Related issues:

- [https://github.com/keras-team/keras/issues/4506](https://github.com/keras-team/keras/issues/4506)
- [https://stackoverflow.com/questions/63068206/keras-metric-for-multiple-outputs](https://stackoverflow.com/questions/63068206/keras-metric-for-multiple-outputs)
- [https://stackoverflow.com/questions/63058307/tensorflow-dataset-mask-sequence-for-evaluation](https://stackoverflow.com/questions/63058307/tensorflow-dataset-mask-sequence-for-evaluation)
- [https://datascience.stackexchange.com/questions/54443/keras-custom-metric-function-how-to-feed-2-model-outputs-to-a-single-metric-eval](https://datascience.stackexchange.com/questions/54443/keras-custom-metric-function-how-to-feed-2-model-outputs-to-a-single-metric-eval)

**Comment by AndreasMadsen** [comment]
Just wanted to echo this!

It is currently not possible to have a metric that aggregates multiple model-outputs, without resorting to callback hacks. A common real-world example is the Labeled Attachment Score (LAS), often used in NLP Dependency Parsing. A model essentially outputs both an index and a label, and the LAS is essentially an accuracy measure for when both matches.

**Comment by PowerToThePeople111** [comment]
Another small note of interest for this. 

We are building a model that is learning to solve several very similar tasks: the kind of task is a feature of the dataset. Instead of evaluating the models against a full dataset, we want to evaluate it separately against each groups of tasks in a dataset:
So instead of having an AUC for all elements of the dataset, we want to calculate for example 3 different AUCs in case the dataset contains 3 different types of tasks. The final result should be just the average of those 3 AUCs.

You could argue, that after training the model, that I can evaluate it exactly like this (which is true), BUT the problem comes when doing Hyperparameter Tuning where you need to provide a single metric that handles the dataset in the described way. Since ... 

1. one of the features is telling the model which kind of task it shall do AND
2. metrics can not access input features

... I can not solve that task rn.

It would be nice if a solution could be found.

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.

**Comment by AmitMY** [comment]
if you don't want to check if this is still an issue, i certainly do not want to.
